x-pack/plugins/endpoint/server/routes/resolver/utils/pagination.ts

```diff
@@ -56,36 +63,45 @@ export function getPaginationParams(limit: number, after?: string): PaginationPa
   return { size: limit };
 }
 
-export function paginate(pagination: PaginationParams, field: string, query: JsonObject) {
+export function paginate(
+  pagination: PaginationParams,
+  tiebreaker: string,
+  aggregator: string,
+  query: JsonObject
+) {
   const { size, timestamp, eventID } = pagination;
-  query.sort = [{ '@timestamp': 'asc' }, { [field]: 'asc' }];
-  query.aggs = { total: { value_count: { field } } };
+  query.sort = [{ '@timestamp': 'asc' }, { [tiebreaker]: 'asc' }];
+  query.aggs = query.aggs || {};
+  query.aggs = Object.assign({}, query.aggs, { totals: { terms: { field: aggregator, size } } });
   query.size = size;
   if (timestamp && eventID) {
     query.search_after = [timestamp.getTime(), eventID] as Array<number | string>;
   }
   return query;
 }
 
-export function paginatedResults(
-  response: SearchResponse<ResolverEvent>
-): { total: number; results: ResolverEvent[]; nextCursor: string | null } {
-  const total = response.aggregations?.total?.value || 0;
-  if (response.hits.hits.length === 0) {
-    return { total, results: [], nextCursor: null };
+export function buildPaginationCursor(total: number, results: ResolverEvent[]): string | null {
+  if (total > results.length && results.length > 0) {
+    const lastResult = results[results.length - 1];
+    const cursor = {
+      timestamp: lastResult['@timestamp'],
+      eventID: extractEventID(lastResult),
+    };
+    return urlEncodeCursor(cursor);
   }
+  return null;
+}
 
-  const results: ResolverEvent[] = [];
-  for (const hit of response.hits.hits) {
-    results.push(hit._source);
+export function paginatedResults(response: SearchResponse<ResolverEvent>): PaginatedResults {
+  if (response.hits.hits.length === 0) {
+    return { totals: {}, results: [] };
   }
 
-  // results will be at least 1 because of length check at the top of the function
-  const next = results[results.length - 1];
-  const cursor = {
-    timestamp: next['@timestamp'],
-    eventID: extractEventID(next),
-  };
+  const totals = response.aggregations?.totals?.buckets?.reduce(
+    (cummulative: any, bucket: any) => ({ ...cummulative, [bucket.key]: bucket.doc_count }),
```

Comment:
So, TBH, the type checker gets us nothing here. The type for `aggregations` is defined as `any` in the elasticsearch typings:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fb3c5606ad782e7666febfd1d3f4829c15c4f92e/types/elasticsearch/index.d.ts#L629

So... frankly, this would just be semantics at this point, and wouldn't guarantee anything. If we changed the aggregation under the hood such that it wasn't bucketed (like for a `filter` aggregation), or if we used something that didn't have the same structure, we'd still throw a runtime error.